File: android/Memoling/src/app/memoling/android/adapter/WordAdapter.java
package app.memoling.android.adapter;

import java.util.ArrayList;

import org.json.JSONException;
import org.json.JSONObject;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import app.memoling.android.db.DatabaseHelper;
import app.memoling.android.db.SqliteAdapter;
import app.memoling.android.entity.Language;
import app.memoling.android.entity.SyncAction;
import app.memoling.android.entity.Word;
import app.memoling.android.sync.cloud.ISyncAdapter;
import app.memoling.android.sync.cloud.ISyncEntity;

public class WordAdapter extends SqliteAdapter implements ISyncAdapter {

	public final static String TableName = "Words";

	public WordAdapter(Context context) {
		super(context);
	}

	public static Word bindWord(Cursor cursor) {
		return WordAdapter.bindWord(cursor, "");
	}
	
	public static Word bindWord(Cursor cursor, String prefix) {
		Word word = new Word();
		
		word.setWordId(DatabaseHelper.getString(cursor, prefix + "WordId"));
		word.setDescription(DatabaseHelper.getString(cursor, prefix + "Description"));
		word.setLanguage(Language.parse(DatabaseHelper.getString(cursor, prefix + "LanguageIso639")));
		word.setWord(DatabaseHelper.getString(cursor, prefix + "Word"));
		
		return word;
	}
	
	private static ContentValues createValues(Word word) {
		ContentValues values = new ContentValues();
		values.put("WordId", word.getWordId());
		values.put("Word", word.getWord());
		values.put("LanguageIso639", word.getLanguage().getCode());
		values.put("Description", word.getDescription());
		return values;
	}

	public Word get(String wordId) {
		try {
			return WordAdapter.get(getDatabase(), wordId);
		} finally {
			closeDatabase();
		}
	}
	
	public void insert(Word word, String syncClientId) throws SQLiteException {
		try {
			WordAdapter.insert(getDatabase(), word, syncClientId);
		} finally {
			closeDatabase();
		}
	}
	
	public void update(Word word, String syncClientId) throws SQLiteException {
		try {
			WordAdapter.update(getDatabase(), word, syncClientId);
		} finally {
			closeDatabase();
		}
	}
	
	public void delete(String wordId, String syncClientId) throws SQLiteException {
		try {
			WordAdapter.delete(getDatabase(), wordId, syncClientId);
		} finally {
			closeDatabase();
		}
	}
	
	public static Word get(SQLiteDatabase db, String wordId) {
		String sql = "SELECT WordId, Word, LanguageIso639, Description FROM Words WHERE WordId = ?";
		Cursor cursor = null;
		
		try {
			cursor = db.rawQuery(sql, new String[] { wordId });
			if(cursor.moveToNext()) {
				Word word = WordAdapter.bindWord(cursor);
				return word;
			}
		} finally {
			if(cursor != null) {
				cursor.close();
			}
		}
		
		return null;
	}
	
	public static void insert(SQLiteDatabase db, Word word, String syncClientId) throws SQLiteException {
		SyncAction syncAction = new SyncAction();
		syncAction.setAction(SyncAction.ACTION_INSERT);
		syncAction.setPrimaryKey(word.getWordId());
		syncAction.setTable(TableName);
		syncAction.setSyncClientId(syncClientId);
		
		WordAdapter.insertDbSync(db, word, syncAction);
	}

	public static void update(SQLiteDatabase db, Word word, String syncClientId) throws SQLiteException {
		
		ArrayList<String> updateColumns = new ArrayList<String>();
		Word dbWord = WordAdapter.get(db, word.getWordId());
		
		if(!dbWord.getWord().equals(word.getWord())) {
			updateColumns.add("Word");
		}
		
		if(!dbWord.getDescription().equals(word.getDescription())) {
			updateColumns.add("Description");
		}
		
		if(!dbWord.getLanguage().equals(word.getLanguage())) {
			updateColumns.add("LanguageIso639");
		}
		
		for(String updateColumn : updateColumns) {
			SyncAction syncAction = new SyncAction();
			syncAction.setAction(SyncAction.ACTION_UPDATE);
			syncAction.setPrimaryKey(word.getWordId());
			syncAction.setTable(TableName);
			syncAction.setSyncClientId(syncClientId);
			syncAction.setUpdateColumn(updateColumn);
			
			WordAdapter.updateDbSync(db, word, syncAction);
		}
		
	}
	
	public static void delete(SQLiteDatabase db, String wordId, String syncClientId) throws SQLiteException {
		SyncAction syncAction = new SyncAction();
		syncAction.setAction(SyncAction.ACTION_DELETE);
		syncAction.setPrimaryKey(wordId);
		syncAction.setTable(TableName);
		syncAction.setSyncClientId(syncClientId);
		
		WordAdapter.deleteDbSync(db, wordId, syncAction);
	}
	
	@Override
	public ISyncEntity decodeEntity(JSONObject json) throws JSONException {
		Word word = new Word();
		word.decodeEntity(json);		
		return word;
	}

	@Override
	public ISyncEntity getEntity(String primaryKey) {
		return WordAdapter.get(this.getDatabase(), primaryKey);
	}

	@Override
	public void insertEntity(SQLiteDatabase db, ISyncEntity object, SyncAction syncAction) throws SQLiteException {
		WordAdapter.insertDbSync(db, (Word)object, syncAction);
	}

	@Override
	public void deleteEntity(SQLiteDatabase db, String primaryKey, SyncAction syncAction) throws SQLiteException {
		WordAdapter.deleteDbSync(db, primaryKey, syncAction);
	}

	@Override
	public void updateEntity(SQLiteDatabase db, ISyncEntity object, SyncAction syncAction) throws SQLiteException {
		WordAdapter.updateDbSync(db, (Word)object, syncAction);
	}
	
	@Override
	public void buildInitialSync(SQLiteDatabase db, String syncClientId) {
		String sql = "SELECT WordId FROM Words";
		
		Cursor cursor = null;
		try {
			cursor = db.rawQuery(sql, null);
			while(cursor.moveToNext()) {
				SyncAction syncAction = new SyncAction();
				syncAction.setAction(SyncAction.ACTION_INSERT);
				syncAction.setTable(TableName);
				syncAction.setPrimaryKey(DatabaseHelper.getString(cursor, "WordId"));
				syncAction.setSyncClientId(syncClientId);
				
				SyncActionAdapter.insert(db, syncAction);
			}
		} finally {
			if(cursor != null) {
				cursor.close();
			}
		}
	}
	
	private static void insertDbSync(SQLiteDatabase db, Word word, SyncAction syncAction) throws SQLiteException {
		db.beginTransaction();

		try {
			ContentValues values = createValues(word);
			
			db.insertOrThrow(TableName, null, values);
			SyncActionAdapter.insertAction(db, syncAction);
			
			db.setTransactionSuccessful();
		} finally {
			db.endTransaction();
		}
	}
	
	private static void updateDbSync(SQLiteDatabase db, Word word, SyncAction syncAction) throws SQLiteException {
		db.beginTransaction();
		
		try {
			ContentValues values = new ContentValues();
			String column = syncAction.getUpdateColumn();
			
			if(column.equals("Word")) {
				values.put("Word", word.getWord());
			} else if(column.equals("LanguageIso639")) {
				values.put("LanguageISo639", word.getLanguage().getCode());
			} else if(column.equals("Description")) {
				values.put("Description", word.getDescription());
			} 

			if(values.size() > 0) {
				db.update(TableName, values, "WordId = ?", new String[] { word.getWordId() });
				SyncActionAdapter.updateAction(db, syncAction);
			}
			
			db.setTransactionSuccessful();
		} finally {
			db.endTransaction();
		}
	}
	
	private static void deleteDbSync(SQLiteDatabase db, String wordId, SyncAction syncAction) throws SQLiteException {
		db.beginTransaction();
		
		try {

			db.delete(TableName, "WordId = ?", new String[] { wordId });
			SyncActionAdapter.deleteAction(db, syncAction);
			
			db.setTransactionSuccessful();
		} finally {
			db.endTransaction();
		}
	}
}

File: www/v2/adapter/WordAdapter.php
<?php

class WordAdapter extends DbAdapter implements SyncAdapter {

	const SYNC_TABLE = "Words";
	
	public function get($wordId) {
		return self::getDb(parent::connect(), $wordId);
	}

	public static function getDb($db, $wordId) {

		$stm = $db->prepare("SELECT * FROM memoling_Words WHERE WordId = :WordId");
		$stm->bindParam("WordId", $wordId);
		$stm->execute();

		if($row = $stm->fetch()) {
			$obj = self::bindWord($row);
			return $obj;
		}

		return null;
	}

	public function getAll($memoBaseId = null) {
		return self::getAllDb(parent::connect(), $memoBaseId);
	}

	public static function getAllDb($db, $memoBaseId = null) {

		if($memoBaseId == null) {
			$stm = $db->prepare("SELECT * FROM memoling_Words");
		} else {
			$stm = $db->prepare("SELECT * FROM memoling_Words WHERE MemoBaseId = :MemoBaseId");
			$stm->bindParam("MemoBaseId", $memoBaseId);
		}

		$stm->execute();

		$list = array();
		while($row = $stm->fetch()) {
			$obj = self::bindWord($row);
			$list[] = $obj;
		}

		return $list;
	}
	
	public static function bindWord($row, $prefix="") {
		$obj = new Word();
		$obj->LanguageIso639 = $row[$prefix."LanguageIso639"];
		$obj->Word = $row[$prefix."Word"];
		$obj->WordId = $row[$prefix."WordId"];
		$obj->Class = $row[$prefix."Class"];
		$obj->Phonetic = $row[$prefix."Phonetic"];
			
		$obj->Description = $row[$prefix."Description"];
		return $obj;
	}

	public function insert($word, $syncClientId=null) {
		return self::inserDb(parent::transConnect(), $word, $syncClientId);
	}

	public function update($word, $syncClientId=null) {
		return self::updateDb(parent::transConnect(), $word, $syncClientId);
	}

	public function delete($wordId, $syncClientId=null) {
		self::deleteDb(parent::transConnect(), $wordId, $syncClientId);
	}
	
	public static function insertDb($dbTran, $word, $syncClientId=null) {

		$syncAction = new SyncAction();
		$syncAction->Action = SyncAction::ACTION_INSERT;
		$syncAction->PrimaryKey = $word->WordId;
		$syncAction->Table = self::SYNC_TABLE;
		$syncAction->SyncClientId = $syncClientId;
		
		return self::insertDbSync($dbTran, $word, $syncAction);
	}

	public static function updateDb($dbTran, $word, $syncClientId=null) {

		$updateColumns = array();
        // This is the case when data is modified by user not sync mechanism
        // Thats why we need to see what have been changed/

        $dbWord = self::getDb($dbTran, $word->WordId);

        if($dbWord->Word != $word->Word) {
            $updateColumns[] = "Word";
        }

        if($dbWord->LanguageIso639 != $word->LanguageIso639) {
            $updateColumns[] = "LanguageIso639";
        }

        if($dbWord->Description != $word->Description) {
            $updateColumns[] = "Description";
        }

        if($dbWord->Class != $word->Class) {
            $updateColumns[] = "Class";
        }

        if($dbWord->Phonetic != $word->Phonetic) {
            $updateColumns[] = "Phonetic";
        }
		
		foreach($updateColumns as $updateColumn) {

			$syncAction = new SyncAction();
			$syncAction->Action = SyncAction::ACTION_UPDATE;
			$syncAction->PrimaryKey = $word->WordId;
			$syncAction->Table = self::SYNC_TABLE;
			$syncAction->SyncClientId = $syncClientId;
			$syncAction->UpdateColumn = $updateColumn;
			
			self::updateDbSync($dbTran, $word, $syncAction);
		}
	}
	
	public static function deleteDb($dbTran, $wordId, $syncClientId=null) {

		$syncAction = new SyncAction();
		$syncAction->Action = SyncAction::ACTION_DELETE;
		$syncAction->PrimaryKey = $wordId;
		$syncAction->Table = self::SYNC_TABLE;
		$syncAction->SyncClientId = $syncClientId;
		
		return self::deleteDbSync($dbTran, $wordId, $syncAction);
	}

	public function decodeEntity($json) {
		$word = new Word();
		$word->decode($json);
		return $word;
	}
	
	public function getEntity($primaryKey) {
		$obj = $this->get($primaryKey);
		return $obj;
	}
	
	public function insertEntity($dbTran, $object, $syncAction) {
		self::insertDbSync($dbTran, $object, $syncAction);
	}

	public function updateEntity($dbTran, $object, $syncAction) {
		self::updateDbSync($dbTran, $object, $syncAction);
	}
	
	public function deleteEntity($dbTran, $primaryKey, $syncAction) {
		self::deleteDbSync($dbTran, $primaryKey, $syncAction);
	}
	
	private static function insertDbSync($dbTran, $word, $syncAction) {
		$dbTran->beginTransaction();
		
		$query = "INSERT INTO
				memoling_Words (WordId, LanguageIso639, Word, Description)
				VALUES(:Wid,:Lang,:Word,:Description)";
		
		$stm = $dbTran->prepare($query);
		$stm->bindParam("Wid", $word->WordId);
		$stm->bindParam("Lang", $word->LanguageIso639);
		$stm->bindParam("Word", $word->Word);
		$stm->bindParam("Description", $word->Description);
		
		if(!$stm->execute()) {
			$dbTran->rollBack();
			throw new SqlException("Failed to create new Word", $dbTran);
		}
		
		SyncActionAdapter::insertAction($dbTran, $syncAction);

		$dbTran->commit();
	}
	
	private static function updateDbSync($dbTran, $word, $syncAction) {
		$dbTran->beginTransaction();
		
		$query = "UPDATE memoling_Words SET ";
		$value = null;
		$type = null;
		
		switch($syncAction->UpdateColumn) {
			case "Word":
				$query .= "Word = :Value ";
				$value = $word->Word;
				$type = PDO::PARAM_STR;
				break;
			case "LanguageIso639":
				$query .= "LanguageIso639 = :Value ";
				$value = $word->LanguageIso639;
				$type = PDO::PARAM_STR;
				break;
			case "Description":
				$query .= "Description = :Value ";
				$value = $word->Description;
				$type = PDO::PARAM_STR;
				break;
			case "Phonetic":
				$query .= "Phonetic = :Value ";
				$value = $word->Phonetic;
				$type = PDO::PARAM_STR;
				break;
			case "Class";
				$query .= "Class = :Value ";
				$value = $word->Class;
				$type = PDO::PARAM_STR;
				break;
		}
		
		$query .= "WHERE WordId = :WordId";
		
		$stm = $dbTran->prepare($query);
		$stm->bindParam("WordId", $word->WordId);
		
		if($syncAction->UpdateColumn != null)
		$stm->bindParam("Value", $value, $type);

		if(!$stm->execute()) {
			$dbTran->rollBack();
			throw new SqlException("Failed to update Word", $dbTran);
		}
		
		SyncActionAdapter::updateAction($dbTran, $syncAction);

		$dbTran->commit();
	}
	
	
	private static function deleteDbSync($dbTran, $wordId, $syncAction) {
		$dbTran->beginTransaction();
		
		$query = "DELETE FROM
				memoling_Words
				WHERE
				WordId = :WordId
				";
		
		$stm = $dbTran->prepare($query);
		$stm->bindParam(":WordId", $wordId);
		if(!$stm->execute()) {
			$dbTran->rollBack();
			throw new SqlException("Failed to delete Word", $dbTran);
		}
			
		SyncActionAdapter::deleteAction($dbTran, $syncAction);
		
		$dbTran->commit();
	}
}

?>